sql/rpl_parallel.cc

```diff
@@ -3050,13 +3065,12 @@ rpl_parallel::stop_during_until()
 }
 
 
-bool
-rpl_parallel::workers_idle(Relay_log_info *rli)
+void rpl_parallel::update_workers_idle(Relay_log_info *rli)
 {
   mysql_mutex_assert_owner(&rli->data_lock);
-  return !rli->last_inuse_relaylog ||
-    rli->last_inuse_relaylog->queued_count ==
-    rli->last_inuse_relaylog->dequeued_count;
```

Comment:
This previous iteration checked null, but when does `Relay_log_info::last_inuse_relaylog` become `nullptr`?

And how could the old code only check the last in-use relay log rather than all of them?
Wouldn’t the last one have a blank queue after a relay log switch?

Reply:
A couple things cause `last_inuse_relaylog` to nullify:
 1. At SQL thread start, there is a brief moment in-between `reset_inuse_relaylog()` and `alloc_inuse_relaylog()` where this was NULL. Prior to 13a9395a6acf1bd90370bd6fe7b4637597b921dd, this section was not locked, and `SHOW SLAVE STATUS` would result in a segfault in this point.
 2. When the SQL shuts down
 3. When the `Master_info` parent object is destroyed (e.g. `RESET SLAVE` or `CHANGE MASTER TO`). Note the `Relay_log_info` member variable is not a reference, but belongs within the memory of the `Master_info`.

Reply:
And to

> And how could the old code only check the last in-use relay log rather than all of them?
> Wouldn’t the last one have a blank queue after a relay log switch?

Hmm good observation, there may indeed be a small gap of time right after rotating, and before the new binlog's `Format_description_event` is queued where we temporarily appear up-to-date. Can you test that? I think your patch would still be vulnerable to this.